### PR TITLE
OSL surface area

### PR DESCRIPTION
**Pull request by est77**
surfacearea() can now be used in OSL shaders that have emission closures.
Because of hierarchical instancing, the surface area cannot be stored in the objects themselves.
Instead, it's stored in the ShaderGroups.

I'll add a test scene when I can run appleseed.studio (it does not compile for me at the moment).

For some reason I cannot rebase this PR. Git shows 1 commit in my local repo but when I push it 
I get 4.

### CODE REVIEW

File: src/appleseed/renderer/kernel/lighting/lightsampler.cpp [hunk 1]
```diff
@@ -238,6 +268,8 @@ void LightSampler::collect_emitting_triangles(
         // Retrieve the object instance.
         const ObjectInstance* object_instance = assembly.object_instances().get_by_index(object_instance_index);
 
+        double object_area = 0.0;
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/lighting/lightsampler.cpp]: Move after `if` below.

File: src/appleseed/renderer/modeling/shadergroup/shadergroup.h [hunk 2]
```diff
@@ -115,13 +118,18 @@ class APPLESEED_DLLSYMBOL ShaderGroup
     // Return true if the shader group contains at least one debug closure.
     bool has_debug() const;
 
-    // Return true if the shader group uses dPdtime.
+    // Return true if the shader group uses the dPdtime global.
     bool uses_dPdtime() const;
 
+    // Return the surface area of an object.
+    // Can only be called if the shader group has emission closures.
+    double surface_area(const AssemblyInstance* ass, const ObjectInstance* obj) const;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/shadergroup/shadergroup.h]: `get_surface_area()`

File: src/appleseed/renderer/modeling/shadergroup/shadergroup.cpp [hunk 3]
```diff
@@ -67,9 +71,13 @@ namespace
 
 struct ShaderGroup::Impl
 {
-    ShaderContainer                 m_shaders;
-    ShaderConnectionContainer       m_connections;
-    mutable OSL::ShaderGroupRef     m_shader_group_ref;
+    typedef pair<const AssemblyInstance*,const ObjectInstance*> SurfaceAreaKey;
+    typedef boost::unordered_map<SurfaceAreaKey,double>         SurfaceAreaMap;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/shadergroup/shadergroup.cpp]: Add space after comma.

File: src/appleseed/renderer/modeling/shadergroup/shadergroup.cpp [hunk 4]
```diff
@@ -41,8 +41,12 @@
 #include "foundation/utility/foreach.h"
 #include "foundation/utility/uid.h"
 
+// boost headers
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/shadergroup/shadergroup.cpp]: `// Boost headers.`

File: src/appleseed/renderer/kernel/shading/shadingpoint.cpp [hunk 5]
```diff
@@ -796,7 +795,14 @@ void ShadingPoint::initialize_osl_shader_globals(
 
     // Always update the ray type and surface area.
     m_shader_globals.raytype = static_cast<int>(ray_flags);
-    m_shader_globals.surfacearea = surface_area;
+
+    if (ray_flags == VisibilityFlags::LightRay && sg.has_emission())
+    {
+        m_shader_globals.surfacearea =
+            static_cast<float>(sg.surface_area( &get_assembly_instance(), &get_object_instance()));
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/shadingpoint.cpp]: Remove space after parenthesis.

**Comment by dictoon** [comment]
A few minor things to fix, looks good to me otherwise!

**Comment by est77** [comment]
Fixed and merged

**Pull request merged**